### PR TITLE
Uplifting 8051 architecture to new IL

### PR DESCRIPTION
**Pull request by b1llow**
# SQUASH ME

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

continue https://github.com/rizinorg/rizin/pull/1609

Instrunction set: https://www.win.tue.nl/~aeb/comp/8051/set8051.html

/ | 0x00 | 0x01 | 0x02 | 0x03 | 0x04 | 0x05 | 0x06 | 0x07 | 0x08 | 0x09 | 0x0A | 0x0B | 0x0C | 0x0D | 0x0E | 0x0F
 --| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0x00|NOP | AJMP | LJMP | RR | INC | INC | INC | INC | INC | INC | INC | INC | INC | INC | INC | INC
0x10|JBC | ACALL | LCALL | RRC | DEC | DEC | DEC | DEC | DEC | DEC | DEC | DEC | DEC | DEC | DEC | DEC
0x20|JB | AJMP | RET | RL | ADD | ADD | ADD | ADD | ADD | ADD | ADD | ADD | ADD | ADD | ADD | ADD
0x30|JNB | ACALL | RETI | RLC | ADDC | ADDC | ADDC | ADDC | ADDC | ADDC | ADDC | ADDC | ADDC | ADDC | ADDC | ADDC
0x40|JC | AJMP | ORL | ORL | ORL | ORL | ORL | ORL | ORL | ORL | ORL | ORL | ORL | ORL | ORL | ORL
0x50|JNC | ACALL | ANL | ANL | ANL | ANL | ANL | ANL | ANL | ANL | ANL | ANL | ANL | ANL | ANL | ANL
0x60|JZ | AJMP | XRL | XRL | XRL | XRL | XRL | XRL | XRL | XRL | XRL | XRL | XRL | XRL | XRL | XRL
0x70|JNZ | ACALL | ORL | JMP | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV
0x80|SJMP | AJMP | ANL | MOVC | DIV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV
0x90|MOV | ACALL | MOV | MOVC | SUBB | SUBB | SUBB | SUBB | SUBB | SUBB | SUBB | SUBB | SUBB | SUBB | SUBB | SUBB
0xA0|ORL | AJMP | MOV | INC | MUL | ? | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV
0xB0|ANL | ACALL | CPL | CPL | CJNE | CJNE | CJNE | CJNE | CJNE | CJNE | CJNE | CJNE | CJNE | CJNE | CJNE | CJNE
0xC0|PUSH | AJMP | CLR | CLR | SWAP | XCH | XCH | XCH | XCH | XCH | XCH | XCH | XCH | XCH | XCH | XCH
0xD0|POP | ACALL | SETB | SETB | DA | DJNZ | XCHD | XCHD | DJNZ | DJNZ | DJNZ | DJNZ | DJNZ | DJNZ | DJNZ | DJNZ
0xE0|MOVX | AJMP | MOVX | MOVX | CLR | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV
0xF0|MOVX | ACALL | MOVX | MOVX | CPL | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV | MOV

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by XVilka** [comment]
Please add also yourself in SPDX copyright section of that file and @Basstorm as well.

File: librz/analysis/arch/8051/8051_il.c [hunk 1]
```diff
@@ -0,0 +1,461 @@
+// SPDX-FileCopyrightText: 2021 Basstorm <?>
```
Comment (XVilka) [inline librz/analysis/arch/8051/8051_il.c]: @Basstorm what should we put here?

Comment (wargio) [reply librz/analysis/arch/8051/8051_il.c]: `// SPDX-FileCopyrightText: 2021 Basstorm <basstorm@nyist.edu.cn>`

Comment (Basstorm) [reply librz/analysis/arch/8051/8051_il.c]: My SPDX copyright information can be completely removed, because my previous PR has no reference value.😅

**Comment by b1llow** [comment]
The format error seems to be because the clang-format version in my system is 15

File: librz/analysis/arch/8051/8051_il.c [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+	rz_return_val_if_fail(op && op->argv[0] && op->argv[1], NULL);
+	if (op->argv[0]->mode == I8051_ADDRESSING_BIT || op->argv[1]->mode == I8051_ADDRESSING_BIT) {
+		return set_any(op->argv[0], OR(get_any(op->argv[0]), get_any(op->argv[1])));
+	}
+	return set_any(op->argv[0], LOGOR(get_any(op->argv[0]), get_any(op->argv[1])));
+}
+static RzILOpEffect *i_xrl(I8051Op *op) {
+	rz_return_val_if_fail(op && op->argv[0] && op->argv[1], NULL);
+	if (op->argv[0]->mode == I8051_ADDRESSING_BIT || op->argv[1]->mode == I8051_ADDRESSING_BIT) {
+		return set_any(op->argv[0], XOR(get_any(op->argv[0]), get_any(op->argv[1])));
```
Comment (thestr4ng3r) [inline librz/analysis/arch/8051/8051_il.c]: Please add an asm test to cover this case.

File: librz/analysis/meson.build [hunk 3]
```diff
@@ -276,6 +276,7 @@ rz_analysis = library('rz_analysis', rz_analysis_sources,
     rz_bin_dep,
     rz_type_dep,
     rz_il_dep,
+    rz_io_dep,
```
Comment (thestr4ng3r) [inline librz/analysis/meson.build]: Why is this added here?

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]
Looks very nice already!

File: librz/analysis/meson.build (hunk 3 above)
Comment (b1llow) [reply librz/analysis/meson.build]: Now it is useless

File: librz/analysis/arch/8051/8051_il.c (hunk 2 above)
Comment (b1llow) [reply librz/analysis/arch/8051/8051_il.c]: I checked and found that this case does not happen

File: librz/analysis/arch/8051/8051_il.c (hunk 1 above)
Comment (XVilka) [reply librz/analysis/arch/8051/8051_il.c]: Removing is fine too if you want. @imbillow

**Review by thestr4ng3r — APPROVED** [review]
Great job!

File: librz/analysis/arch/8051/8051_il.c [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+}
+
+static RzILOpPure *carryout(RzILOpBitVector *a, RzILOpBitVector *b, RzILOpBitVector *c) {
+	return LOGOR(LOGAND(a, b), LOGAND(NEG(c), LOGOR(DUP(a), DUP(b))));
+}
+
+static RzILOpEffect *set_add_carry(RzILOpBitVector *a, RzILOpBitVector *b, RzILOpBitVector *c) {
+	rz_return_val_if_fail(a && b && c, NULL);
+	RzILOpPure *ov = OR(AND3(MSB(a), MSB(b), INV(MSB(c))), AND3(INV(MSB(DUP(a))), INV(MSB(DUP(b))), MSB(DUP(c))));
+	//	return set_psw_carry_flags(BOOL_TO_BV(cy, 8), BOOL_TO_BV(ac, 8), BOOL_TO_BV(ov, 8));
```
Comment (XVilka) [inline librz/analysis/arch/8051/8051_il.c]: If it's not used - remove

**Review by wargio — APPROVED** [review]

File: librz/analysis/arch/8051/8051_il.c [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+#define VAL_A        VARG_8051(I8051_A)
+#define VAL_B        VARG_8051(I8051_B)
+#define VAL_CY       VARG_8051(I8051_CY)
+#define VAL_CY8      BOOL_TO_BV(VARG_8051(I8051_CY), 8)
+
+static RzILOpPure *val_register(I8051OpAddressing *a) {
+	switch (a->d.reg) {
+	case I8051_PC:
+		return U16(a->pc);
+		//	case I8051_CY:
```
Comment (XVilka) [inline librz/analysis/arch/8051/8051_il.c]: Why is this commented out?

**Review by XVilka — APPROVED** [review]
Looks good, apart from a couple of comments. The remaining part would be the tracetest.

File: librz/analysis/arch/8051/8051_il.c [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+}
+
+static RzILOpPure *get_any(I8051OpAddressing *a) {
+	switch (a->mode) {
+	case I8051_ADDRESSING_REGISTER:
+		return val_register(a);
+	case I8051_ADDRESSING_DIRECT:
+		return LOAD(U16(a->d.addr));
+	case I8051_ADDRESSING_INDIRECT:
+		return LOAD(CAST(16, IL_FALSE, get_any(a->d.indirect)));
```
Comment (DMaroo) [inline librz/analysis/arch/8051/8051_il.c]: You could just use `UNSIGNED(16, get_any(a->d.indirect))` here. It is more concise and readable.

File: librz/analysis/arch/8051/8051_il.c [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+	case I8051_ADDRESSING_IMMEDIATE:
+		return U8(a->d.constant & 0xff);
+	case I8051_ADDRESSING_IMMEDIATE16:
+		return U16(a->d.constant);
+	case I8051_ADDRESSING_RELATIVE:
+		return ADD(U16(a->pc), U16(a->d.constant));
```
Comment (DMaroo) [inline librz/analysis/arch/8051/8051_il.c]: You can just have `U16(a->pc + a->d.constant)` instead. I am not saying either one is better, just saying that this would make the IL concise.

File: librz/analysis/arch/8051/8051_il.c (hunk 6 above)
Comment (DMaroo) [reply librz/analysis/arch/8051/8051_il.c]: Similarly for the rest of the occurrences of `CAST`.

File: librz/analysis/arch/8051/8051_il.c [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+}
+static RzILOpEffect *set_any(I8051OpAddressing *a, RzILOpPure *v) {
+	switch (a->mode) {
+	case I8051_ADDRESSING_REGISTER:
+		return set_reg(a->d.reg, v);
+	case I8051_ADDRESSING_BIT: {
+		return STOREW(U16(a->d.addr), BOOL_TO_BV(v, 1));
+	}
+	case I8051_ADDRESSING_DIRECT:
+		return STORE(U16(a->d.addr), v);
```
Comment (DMaroo) [inline librz/analysis/arch/8051/8051_il.c:127]: Instead of duplicating the address calculation logic, just use `return STOREW(get_any(a), v)`.

**Review by DMaroo — CHANGES_REQUESTED** [review]
Looks pretty neat to me. Great work!

File: librz/analysis/arch/8051/8051_il.c [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+	rz_return_val_if_fail(a && b && c, NULL);
+	RzILOpPure *ov = OR(AND3(MSB(a), MSB(b), INV(MSB(c))), AND3(INV(MSB(DUP(a))), INV(MSB(DUP(b))), MSB(DUP(c))));
+	//	return set_psw_carry_flags(BOOL_TO_BV(cy, 8), BOOL_TO_BV(ac, 8), BOOL_TO_BV(ov, 8));
+	return SEQ4(SETL("carry", carryout(DUP(a), DUP(b), DUP(c))),
+		set_reg(I8051_CY, bv_i(VARL("carry"), 7)),
+		set_reg(I8051_AC, bv_i(VARL("carry"), 3)),
+		set_reg(I8051_OV, ov));
+}
+
+static RzILOpPure *borrowin(RzILOpBitVector *a, RzILOpBitVector *b, RzILOpBitVector *c) {
```
Comment (Rot127) [inline librz/analysis/arch/8051/8051_il.c:163]: `rz_return_val_if_fail(a && b && c, NULL);`

File: librz/analysis/arch/8051/8051_il.c [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+}
+
+static RzILOpEffect *i_subb(I8051Op *op) {
+	rz_return_val_if_fail(op && op->argv[0], NULL);
+	return SEQ4(SETL("arg1", get_any(op->argv[1])),
+		SETL("res", SUB3(VAL_A, VARL("arg1"), VAL_CY8)),
+		set_sub_curry(VAL_A, VARL("arg1"), VARL("res")),
+		set_a(VARL("res")));
+}
+static RzILOpEffect *i_div(I8051Op *op) {
```
Comment (Rot127) [inline librz/analysis/arch/8051/8051_il.c:201]:
```suggestion
static RzILOpEffect *i_div(I8051Op *op) {
	rz_return_val_if_fail(op && op->argv[0], NULL);
```

Comment (Rot127) [reply librz/analysis/arch/8051/8051_il.c:201]: And similar cases below.

File: librz/analysis/arch/8051/8051_op.h [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+	I8051_ADDRESSING_REGISTER = 0,
+	I8051_ADDRESSING_DIRECT,
+	I8051_ADDRESSING_INDIRECT,
+	I8051_ADDRESSING_IMMEDIATE,
+	I8051_ADDRESSING_IMMEDIATE16,
+	I8051_ADDRESSING_RELATIVE,
+	I8051_ADDRESSING_ABSOLUTE,
+	I8051_ADDRESSING_LONG,
+	I8051_ADDRESSING_INDEXED,
+
```
Comment (Rot127) [inline librz/analysis/arch/8051/8051_op.h]:
```suggestion
```

File: librz/analysis/arch/8051/8051_op.h [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+static inline I8051OpAddressing *addressing_register_b() {
+	return addressing_register(I8051_B);
+}
+
+/**
+ * 0x05 iram addr
+ * 0x06-0x07 @(R0-R1)
+ * 0x08-0x0f, R0-R7
+ */
```
Comment (Rot127) [inline librz/analysis/arch/8051/8051_op.h]: Is the meaning here common knowledge for people who know 8051? Otherwise I would suggest to add more context (this is the offset into the buffer?).

Comment (Rot127) [reply librz/analysis/arch/8051/8051_op.h]: Same for comments below.

File: librz/analysis/arch/8051/8051_op.h [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+	}
+	RZ_LOG_DEBUG("invalid addressing pattern 1")
+	return NULL;
+}
+
+/**
+ * \brief 0x00 \@DPTR
+ *        0x02-0x03 @(R0|R1)
+ */
+static I8051OpAddressing *addressing_pattern11(const ut8 *buf) {
```
Comment (Rot127) [inline librz/analysis/arch/8051/8051_op.h]: Is there a reason that the function definitions are in the header file?

**Review by Rot127 — APPROVED** [review]
Could only skim over it briefly. But LGTM. Good job!

**Review by Basstorm — APPROVED** [review]

**Review by DMaroo — APPROVED** [review]

File: librz/analysis/arch/8051/8051_il.c (hunk 5 above)
Comment (b1llow) [reply librz/analysis/arch/8051/8051_il.c]: Obsolete codes

File: librz/analysis/arch/8051/8051_il.c (hunk 8 above)
Comment (b1llow) [reply librz/analysis/arch/8051/8051_il.c:127]: But it should not be equivalent

**Review by XVilka — CHANGES_REQUESTED** [review]
@imbillow many tests are broken on big endian platform:
```
[XX] db/asm/8051 <asm> djnz 0x40, 0x118b
-- <asm> djnz 0x40, 0x118b <--> d54088 ---> <IL>
-- assembly
-- IL
--- expected
+++ actual

@@ -1,1 +1,1 @@
-(seq (set res (- (load 0 (bv 16 0x40)) (bv 8 0x1))) (store 0 (bv 16 0x40) (var res)) (branch (! (is_zero (var res))) (jmp (bv 16 0x1288)) nop))
+(seq (set res (- (load 0 (bv 16 0x0)) (bv 8 0x1))) (store 0 (bv 16 0x0) (var res)) (branch (! (is_zero (var res))) (jmp (bv 16 0x1288)) nop))

[XX] db/asm/8051 <asm> dec 0x35
-- <asm> dec 0x35 <--> 1535 ---> <IL>
-- assembly
-- IL
--- expected
+++ actual
@@ -1,1 +1,1 @@
-(seq (set arg0 (load 0 (bv 16 0x35))) (store 0 (bv 16 0x35) (- (var arg0) (bv 8 0x1))))
+(seq (set arg0 (load 0 (bv 16 0x0))) (store 0 (bv 16 0x0) (- (var arg0) (bv 8 0x1))))
```
See all at https://app.travis-ci.com/github/rizinorg/rizin/jobs/590853977

File: librz/analysis/arch/8051/8051_parse.c [hunk 14]
```diff
@@ -392,7 +392,9 @@ RZ_IPI I8051Op *rz_8051_op_parse(const ut8 *buf, ut64 len, ut64 pc) {
 			switch (lo) {
 			case 0x0: {
 				op->argv[0] = addressing_register(I8051_DPTR);
-				op->argv[1] = addressing_addr16(I8051_ADDRESSING_IMMEDIATE16, (buf[1] << 8) | buf[2]);
+				ut16 imm = analysis->big_endian ? (buf[1] | (buf[2] << 8))
```
Comment (XVilka) [inline librz/analysis/arch/8051/8051_parse.c]: @imbillow you could just use `rz_read_ble16()` here instead

**Review by XVilka — APPROVED** [review]
LGTM
